### PR TITLE
Create input pipeline for current TF2 implementation.

### PR DESCRIPTION
**Pull request by boru-roylu**
- Originally, we have input_utils.py and pretrain_utils.py. We rearrange them into two files, input_utils.py and input_pipeline.py. Thus, pretrain_utils.py is deleted.
- In the future, all functions used to create `TF.data.Dataset` will be put in input_pipeline.py.
- Move all functions related to `tf.data.Dataset.map` previously in pretrain_utils.py to the file input_utils.py.
- The function `get_masking_fn` in input_utils.py is mainly from `input_fn_builder` previously in pretrain_utils.py. Thus, lines from 339 to 555, except lines from 504 to 515, are copied from `input_fn_builder`.
- Lines from 504 to 515 are used to zero out patch embeddings that should be masked.
- Rename function `make_masked_language_model_and_masked_patch_prediction_features` to `make_mlm_and_mpp_features`.
- The new function `add_fake_labels` in input_utils.py is adhoc method for current TF2 implementation.

### CODE REVIEW

File: src/input_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                            input_config,
+                            model_config,
+                            batch_size,
+                            tokenizer,
+                            is_training=True,
+                            input_pipeline_context=None,
+                            output_fake_labels=True):
+  """Creates input dataset from (tf)records files for pretraining."""
+
+  name_to_features = {'image_data': tf.io.FixedLenFeature([], tf.string)}
```
Comment (wqfcr-google) [inline src/input_pipeline.py:32]: Nit: do we want to specify the shape here instead of [] (if they are resized to the same size)?

File: src/input_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        functools.partial(input_utils.add_side_input_features,
+                          input_config, model_config),
+        tf.data.experimental.AUTOTUNE)
+
+  if output_fake_labels:
+    dataset = dataset.map(
+        input_utils.add_fake_labels,
+        num_parallel_calls=tf.data.experimental.AUTOTUNE)
+
+
```
Comment (wqfcr-google) [inline src/input_pipeline.py:94]: Nit: remove extra line?

File: src/input_utils.py [hunk 3]
```diff
@@ -55,13 +57,13 @@ class PretrainInputConfig(object):
   # The fraction of tokens to mask for masked patch prediction (mpp) loss.
   mpp_fraction_to_mask = attr.ib(default=0.50)
 
-  # Maximum number of masked text tokens per batch.
-  mlm_max_selections_per_batch = attr.ib(default=480)
+  # Maximum number of masked text tokens per sequence.
+  mlm_max_selections_per_sequence = attr.ib(default=47)
```
Comment (wqfcr-google) [inline src/input_utils.py:61]: Nit: how do you determine the 47 and 98 numbers?

File: src/input_utils.py [hunk 4]
```diff
@@ -315,17 +315,241 @@ def add_side_input_features(
...
 
   features.update(side_inputs.to_dict())
 
-  # TODO (roylu): figure out a better solution.
-  # Add None as dummy label.
-  return features, None
+  return features
+
+
+def add_fake_labels(features):
```
Comment (wqfcr-google) [inline src/input_utils.py:332]: What is the usage of this function? Please add some comments.

**Review by wqfcr-google — COMMENTED** [review]
Some initial comments.

File: src/input_utils.py (hunk 3 above)
Comment (boru-roylu) [reply src/input_utils.py:61]: Originally, I thought the variable `max_selections_per_batch` in [RandomItemSelector](https://github.com/googleinterns/multimodal-long-transformer-2021/blob/roylu-input/src/input_utils.py#L374) is used to determined the maximum number of selected masked tokens per "batch". However, I found that it is per "sentence". Thus, the 98 is from 196 (patch tokens) * 0.5 (patch masking probability) and 48 is from 314 (text tokens) * 0.15 (text masking probability).

Comment (wqfcr-google) [reply src/input_utils.py:61]: I see. Please try to incorporate these into the comment, which help readers to understand these 'magic' numbers.

Comment (boru-roylu) [reply src/input_utils.py:61]: will do!

File: src/input_pipeline.py (hunk 1 above)
Comment (wqfcr-google) [reply src/input_pipeline.py:32]: Or are these the raw image data?

File: src/input_utils.py [hunk 5]
```diff
@@ -315,17 +315,241 @@ def add_side_input_features(
...
+                                        '(b s) h -> b s h',
+                                        b=batch_size)
+  
+    masked_patch_positions = masked_patch_positions.to_tensor()
+    masked_patch_positions = tf.cast(masked_patch_positions, tf.int32)
+    masked_patch_positions = pad_to_max_seq_len(
+        masked_patch_positions, mpp_max_selections_per_sequence)
+    features['masked_patch_positions'] = masked_patch_positions
+                                    
+    # Create label for masked patche prediction.
```
Comment (santiaontanon-google) [inline src/input_utils.py]: "patche" -> "patch"

File: src/input_utils.py [hunk 6]
```diff
@@ -315,17 +315,241 @@ def add_side_input_features(
...
+      A mask that indicate the positions of real predictions.
+
+    """
+    mask_position = tf.equal(masked_token_ids, mask_token_id)
+    mask_position = tf.cast(mask_position, tf.int32)
+    num_real_masked_tokens = tf.reduce_sum(mask_position, axis=1, keepdims=True)
+    position = tf.range(masked_seq_len, dtype=tf.int32)
+    return tf.where(position < num_real_masked_tokens, 1.0, 0.0)
+    
+  def pad_to_max_seq_len(tensor, max_seq_len):
```
Comment (santiaontanon-google) [inline src/input_utils.py:445]: What happens if max_seq_len is actually smaller than the length of tensor?

**Review by santiaontanon-google — COMMENTED** [review]
Thanks Roy!

File: src/input_pipeline.py (hunk 1 above)
Comment (boru-roylu) [reply src/input_pipeline.py:32]: They are raw image data, so the shapes are not always the same.

Comment (wqfcr-google) [reply src/input_pipeline.py:32]: I see. Thanks for the clarification.

**Review by wqfcr-google — APPROVED** [review]

File: src/input_utils.py (hunk 6 above)
Comment (boru-roylu) [reply src/input_utils.py:445]: If we call [get_decode_fn](https://github.com/googleinterns/multimodal-long-transformer-2021/blob/roylu-input/src/input_utils.py#L93) first, we will trim the sequence to be shorter than maximum sequence length. We set make sure that [text part will be shorter than max_text_seq_len](https://github.com/googleinterns/multimodal-long-transformer-2021/blob/roylu-input/src/input_utils.py#L135) and patch part is always 198 (196 + 2). Thus, we will never have a tensor longer than `max_seq_len`.

Comment (santiaontanon-google) [reply src/input_utils.py:445]: ok! If you have a chance, add a comment somewhere to make sure reading the code it's obvious where trimming occurs :)

**Review by santiaontanon-google — APPROVED** [review]
LGTM!

File: src/input_utils.py [hunk 7]
```diff
@@ -17,13 +17,15 @@
 import attr
 import tensorflow as tf
 import tensorflow_text as tf_text
+from einops import reduce, rearrange
 from official.modeling import tf_utils
+from tensorflow.python.ops import math_ops
```
Comment (jtainslie) [inline src/input_utils.py:22]: Nit: not sorted alphabetically.  (Ignore the leading "import"/"from" when sorting.)

File: src/input_utils.py [hunk 8]
```diff
@@ -17,13 +17,15 @@
 import attr
 import tensorflow as tf
 import tensorflow_text as tf_text
+from einops import reduce, rearrange
```
Comment (jtainslie) [inline src/input_utils.py:20]: Nit: imports should be for modules or packages, not individual functions.  https://google.github.io/styleguide/pyguide.html#22-imports

File: src/input_utils.py [hunk 9]
```diff
@@ -315,17 +315,241 @@ def add_side_input_features(
...
+    """Gets mask for real predictions.
+
+    The `positions` tensor might be zero-padded (if the sequence is too short 
+    to have the maximum number of predictions).
+    The `masked_weights` tensor has a value of 1.0 for every real prediction 
+    and 0.0 for the padding predictions.
+
+    Args:
+      masked_token_ids: A tensor that contains mask token ids.
+    Returns:
```
Comment (jtainslie) [inline src/input_utils.py:435]: Nit: there should be a blank space above `Returns:`.

**Review by jtainslie — APPROVED** [review]
Lgtm.
